staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go

```diff
@@ -0,0 +1,276 @@
+	// populate a VersionedOptions with the variable and the type:
+	//
+	//    schema := schemaResolver.ResolveSchema(groupVersionKind)
+	//    objectType := apiservercel.SchemaDeclType(schema, true)
+	//    ...
+	//    VersionOptions{
+	//      IntroducedVersion: version.MajorMinor(1, 26),
+	//      DeclTypes: []*apiservercel.DeclType{ objectType },
+	//      EnvOptions: []cel.EnvOption{ cel.Variable("object", objectType.CelType()) },
+	//    },
+	//
+	// To create an DeclType from a CRD, use a structural schema. For example:
+	//
+	//    schema := structuralschema.NewStructural(crdJSONProps)
+	//    objectType := apiservercel.SchemaDeclType(schema, true)
+	//
+	// Required.
+	IntroducedVersion *version.Version
+	// RemovedVersion is the version at which these options were removed.
+	// The NewExpressions environment will not include options removed at or before the
+	// compatibility version of the EnvSet.
+	//
+	// All option removals must be backward compatible; the removal must either be paired
+	// with a compatible replacement introduced at the same version, or the removal must be non-breaking.
+	// The StoredExpressions environment will not include removed options.
+	//
+	// A function library may be upgraded by setting the RemovedVersion of the old library
+	// to the same value as the IntroducedVersion of the new library. The new library must
+	// be backward compatible with the old library.
+	//
+	// For example:
+	//
+	//    VersionOptions{
+	//      IntroducedVersion: version.MajorMinor(1, 26), RemovedVersion: version.MajorMinor(1, 27),
+	//      EnvOptions: []cel.EnvOption{ libraries.Example(libraries.ExampleVersion(1)) },
+	//    },
+	//    VersionOptions{
+	//      IntroducedVersion: version.MajorMinor(1, 27),
+	//      EnvOptions: []EnvOptions{ libraries.Example(libraries.ExampleVersion(2)) },
+	//    },
+	//
+	// Optional.
+	RemovedVersion *version.Version
+
+	// EnvOptions provides CEL EnvOptions. This may be used to add a cel.Variable, a
+	// cel.Library, or to enable other CEL EnvOptions such as language settings.
+	//
+	// If an added cel.Variable has an OpenAPI type, the type must be included in DeclTypes.
+	EnvOptions []cel.EnvOption
+	// ProgramOptions provides CEL ProgramOptions. This may be used to set a cel.CostLimit,
+	// enable optimizations, and set other program level options that should be enabled
+	// for all programs using this environment.
+	ProgramOptions []cel.ProgramOption
+	// DeclTypes provides OpenAPI type declarations to register with the environment.
+	//
+	// If cel.Variables added to EnvOptions refer to a OpenAPI type, the type must be included in
+	// DeclTypes.
+	DeclTypes []*apiservercel.DeclType
+}
+
+// Extend returns an EnvSet based on this EnvSet but extended with given VersionedOptions.
+// This EnvSet is not mutated.
+// The returned EnvSet has the same compatibility version as the EnvSet that was extended.
+func (e *EnvSet) Extend(options ...VersionedOptions) (*EnvSet, error) {
```

Comment:
I did a quick sweep of calls to this and it looks like the only times we call this per-request is in validation of CRD create/update, which seems fine / slow-moving. Otherwise we're doing this once per update of cel-containing objects. Does that match what you see?

Reply:
do we have any benchmarks that would notice if we accidentally started reconstructing (or re-extending) envsets per-request while validating or admitting? that would help sanity check this rewiring (or future rewirings)

Reply:
Prexisiting:

- CRD Validation - `staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go:BenchmarkCELValidationWithContext` (benchmarks the whole flow that compiles and then runs validations for CRDs)
- CRD Validation - `staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go:BenchmarkCompile` (lower level compile benchmark)

I've added just now:

- base environment `staging/src/k8s.io/apiserver/pkg/cel/environment/base_test.go:MustBaseEnv` (benchmarks the cache)
- admission policies, webhook match conditions - `staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/compile_test.go:BenchmarkCompile` (benchmarks the flow to extend the base env and then validate multiple resources with the compile program)